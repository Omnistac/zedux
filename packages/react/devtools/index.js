if (process.env.NODE_ENV !== 'development') {
  module.exports = {
    getStateHub: function getStateHub() {},
    StateHub: function StateHub() {
      return null
    },
  }
} else {
  module.exports = require('./dev')
}
