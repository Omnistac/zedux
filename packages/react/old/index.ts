export * from './AppProvider'
export { globalStore } from './global/store'
export * from './createAtom'
export * from './createMolecule'
export * from './types'
export * from './useAtom'
export * from './useMolecule'
export * from './useStore'
