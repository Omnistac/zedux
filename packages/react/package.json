{
  "name": "@zedux/react",
  "version": "0.3.4",
  "description": "Elite state management for React",
  "main": "./dist/es/react/src/index.js",
  "module": "./dist/es/react/src/index.js",
  "es2015": "./dist/es/react/src/index.js",
  "publishConfig": {
    "registry": "https://omnistac.jfrog.io/omnistac/api/npm/npm-local/"
  },
  "typings": "./dist/es/react/src/index.d.ts",
  "sideEffects": false,
  "files": [
    "dist"
  ],
  "scripts": {
    "build-storybook": "build-storybook",
    "build:es": "ttsc --outDir dist/es -p tsconfig.es.json",
    "build:umd:min": "cross-env NODE_ENV=production webpack",
    "build:umd": "cross-env NODE_ENV=development webpack",
    "build": "yarn build:es && yarn build:umd && yarn build:umd:min",
    "dev": "yarn build:es -- --watch",
    "prepublishOnly": "yarn build",
    "storybook": "start-storybook -p 6006",
    "test": "cd ../.. && yarn test packages/react"
  },
  "keywords": [
    "abstraction",
    "atom",
    "atoms",
    "atomic",
    "beautiful",
    "boilerplate",
    "DI",
    "dependency",
    "easy",
    "flux",
    "functional",
    "immutable",
    "injection",
    "machine",
    "management",
    "memoized",
    "performant",
    "react",
    "reducer",
    "redux",
    "simple",
    "state",
    "test",
    "tests",
    "testing",
    "zedux"
  ],
  "author": "Joshua Claunch",
  "license": "MIT",
  "repository": {
    "type": "git",
    "url": "git+https://github.com/Omnistac/zedux.git"
  },
  "bugs": {
    "url": "https://github.com/Omnistac/zedux/issues"
  },
  "homepage": "https://github.com/Omnistac/zedux#readme",
  "devDependencies": {
    "@testing-library/jest-dom": "^5.11.9",
    "@testing-library/react": "^11.2.5",
    "@types/styled-components": "^5.1.7",
    "core-js": "3",
    "react": "^17.0.1",
    "react-dom": "^17.0.1",
    "react-icons": "^4.1.0",
    "styled-components": "^5.2.1",
    "ttypescript": "^1.5.12",
    "typescript-transform-paths": "^2.2.3"
  },
  "dependencies": {},
  "peerDependencies": {
    "react": ">=17.0.1",
    "react-dom": ">=17.0.1"
  }
}
