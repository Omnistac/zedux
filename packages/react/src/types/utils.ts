import { AtomTemplateBase } from '../classes/templates/AtomTemplateBase'
import { AtomInstance } from '../classes/instances/AtomInstance'
import { AtomInstanceBase } from '../classes/instances/AtomInstanceBase'

export type AnyAtomInstance = AtomInstance<any>
export type AnyAtomInstanceBase = AtomInstanceBase<
  any,
  AtomTemplateBase<any, AtomInstance<any>>
>
export type AnyAtomTemplate = AtomTemplateBase<any, any>
