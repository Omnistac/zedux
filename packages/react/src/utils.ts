import { AnyAtomTemplate, Ecosystem } from '@zedux/atoms'
import React, { createContext } from 'react'

export const ecosystemContext = createContext('@@global')

/**
 * These are copied from packages/atoms/src/utils/general.ts
 *
 * IMPORTANT: keep these in-sync with the ones in the atoms package
 */
export const Explicit = 1
export const External = 2
export const Static = 4
// export const Deferred = 8

export const destroyed = Symbol.for(`@@zedux/destroyed`)

export const getReactContext = (
  ecosystem: Ecosystem,
  atom: AnyAtomTemplate
) => {
  const reactStorage: {
    contexts: WeakMap<typeof createContext, Record<string, React.Context<any>>>
  } = (ecosystem._storage.react ||= {})

  const contexts = (reactStorage.contexts ||= new WeakMap())
  const windowContexts =
    contexts.get(createContext) ||
    (contexts.set(createContext, {}).get(createContext) as Record<
      string,
      React.Context<any>
    >)

  return (windowContexts[atom.key] ||= createContext(
    undefined
  ) as React.Context<any>)
}
