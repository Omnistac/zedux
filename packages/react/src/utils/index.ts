export * from './general'
export * from './types'
