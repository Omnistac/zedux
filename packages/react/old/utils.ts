import { Atom, Scope } from './types'

let idCounter = 0

const generateId = () => {
  idCounter += 1
  return idCounter
}

const stringifyParam = (param: any): string => {
  if (Array.isArray(param)) {
    return `[${param.map(child => stringifyParam(child)).join(',')}]`
  }

  if (param && typeof param === 'object') {
    return `{${Object.entries(param)
      .map(([key, val]) => `${key}: ${stringifyParam(val)}`)
      .sort((a, b) => a.localeCompare(b))
      .join(',')}}`
  }

  return JSON.stringify(param)
}

export const generateAppId = () => `app-${generateId()}`
export const generateImplementationId = () => `im-${generateId()}`
export const generateInstanceId = () => `in-${generateId()}`
export const generateLocalId = () => `lo-${generateId()}`

export const getFullKey = (atom: Atom, params?: any[]) => {
  if (atom.scope === Scope.local) return generateLocalId()

  if (!params?.length) return atom.key

  return `${atom.key}---${stringifyParam(params)}`
}
