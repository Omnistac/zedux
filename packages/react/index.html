<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="UTF-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Zedux Dev</title>
  </head>
  <body>
    <div id="root"></div>
    <script type="module" src="/src/index.ts"></script>
    <script type="module">
      import { atom, useAtomState } from './src/index'
      import { createElement } from 'react'
      import { render } from 'react-dom'

      const greetingAtom = atom('greeting', 'Hello, World!')

      const App = () => {
        const [state, setState] = useAtomState(greetingAtom)

        return createElement('div', null,
          createElement('p', null, state),
          createElement('input', {
            onChange: event => setState(event.currentTarget.value),
            value: state
          })
        )
      }

      render(createElement(App), document.getElementById('root'))
    </script>
  </body>
</html>
