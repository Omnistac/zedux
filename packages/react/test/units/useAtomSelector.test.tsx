import { AtomGetters, atom, useAtomSelector } from '@zedux/react'
import React, { useState } from 'react'
import { renderInEcosystem } from '../utils/renderInEcosystem'
import {
  ecosystem,
  snapshotNodes,
  snapshotSelectorNodes,
} from '../utils/ecosystem'
import { act, fireEvent } from '@testing-library/react'

describe('useAtomSelector', () => {
  test('triggers a reevaluation on selector instance force-destruction', async () => {
    jest.useFakeTimers()
    const a = { a: 1 }
    const b = { a: 2 }
    let obj = a
    const selector1 = () => obj

    function Test() {
      const { a } = useAtomSelector(selector1)

      return <div data-testid="text">{a}</div>
    }

    const { findByTestId } = renderInEcosystem(<Test />)

    const div = await findByTestId('text')

    expect(div.innerHTML).toBe('1')

    obj = b

    act(() => {
      ecosystem.getNode(selector1, []).destroy(true)
      jest.runAllTimers()
    })

    expect(div.innerHTML).toBe('2')
  })

  test("inline refs are swapped out if the SelectorConfig object's `selector` refs are the same", async () => {
    jest.useFakeTimers()
    const selector1 = jest.fn(() => 'a')

    function Test() {
      const [, setState] = useState(0)
      const val = useAtomSelector({
        selector: selector1,
      })

      return (
        <>
          <button
            data-testid="button"
            onClick={() => setState(state => state + 1)}
          />
          <div data-testid="text">{val}</div>
        </>
      )
    }

    const { findByTestId } = renderInEcosystem(<Test />)

    const button = await findByTestId('button')
    const div = await findByTestId('text')

    expect(div.innerHTML).toBe('a')
    expect(selector1).toHaveBeenCalledTimes(1)

    act(() => {
      fireEvent.click(button)
      jest.runAllTimers()
    })

    expect(div.innerHTML).toBe('a')
    expect(selector1).toHaveBeenCalledTimes(2)
  })

  test('useAtomSelector creates/uses a different instance when args change', async () => {
    jest.useFakeTimers()
    const atom1 = atom('1', (id: number) => id)
    const selector1 = jest.fn(({ get }: AtomGetters, id: number) =>
      get(atom1, [id])
    )

    function Test() {
      const [state, setState] = useState(0)
      const val = useAtomSelector(selector1, state)

      return (
        <>
          <button
            data-testid="button"
            onClick={() => setState(state => state + 1)}
          />
          <div data-testid="text">{val}</div>
        </>
      )
    }

    const { findByTestId } = renderInEcosystem(<Test />)

    const button = await findByTestId('button')
    const div = await findByTestId('text')

    expect(div.innerHTML).toBe('0')
    expect(selector1).toHaveBeenCalledTimes(1)
    expect(ecosystem.dehydrate('@@selector')).toEqual({
      '@@selector-mockConstructor-0-[0]': 0,
    })

    act(() => {
      fireEvent.click(button)
      jest.runAllTimers()
    })

    expect(div.innerHTML).toBe('1')
    expect(selector1).toHaveBeenCalledTimes(2)
    expect(ecosystem.dehydrate('@@selector')).toEqual({
      '@@selector-mockConstructor-0-[1]': 1,
    })
  })

  test('useAtomSelector creates/uses a different instance when selector is actually different', async () => {
    jest.useFakeTimers()
    const atom1 = atom('1', (id: number) => id)
    const selector1 = jest.fn(({ get }: AtomGetters, id: number) =>
      get(atom1, [id])
    )
    const selector2 = jest.fn(() => 2)

    function Test() {
      const [state, setState] = useState(0)
      const val =
        state % 2
          ? useAtomSelector(selector1, state)
          : useAtomSelector(selector2) // yep do it like this

      return (
        <>
          <button
            data-testid="button"
            onClick={() => setState(state => state + 1)}
          />
          <div data-testid="text">{val}</div>
        </>
      )
    }

    const { findByTestId } = renderInEcosystem(<Test />)

    const button = await findByTestId('button')
    const div = await findByTestId('text')

    expect(div.innerHTML).toBe('2')
    expect(selector1).not.toHaveBeenCalled()
    expect(selector2).toHaveBeenCalledTimes(1)
    expect(ecosystem.dehydrate('@@selector')).toEqual({
      '@@selector-mockConstructor-0': 2,
    })

    act(() => {
      fireEvent.click(button)
      jest.runAllTimers()
    })

    expect(div.innerHTML).toBe('1')
    expect(selector1).toHaveBeenCalledTimes(1)
    expect(selector2).toHaveBeenCalledTimes(1)
    expect(ecosystem.dehydrate('@@selector')).toEqual({
      '@@selector-mockConstructor-1-[1]': 1,
    })
  })

  test('useAtomSelector creates/uses a different instance when selector goes from object form to function form', async () => {
    jest.useFakeTimers()
    const selector1 = jest.fn(() => 1)

    function Test() {
      const [state, setState] = useState(0)
      const val = useAtomSelector(
        state % 2 ? selector1 : { selector: selector1 }
      )

      return (
        <>
          <button
            data-testid="button"
            onClick={() => setState(state => state + 1)}
          />
          <div data-testid="text">{val}</div>
        </>
      )
    }

    const { findByTestId } = renderInEcosystem(<Test />)

    const button = await findByTestId('button')
    const div = await findByTestId('text')

    expect(div.innerHTML).toBe('1')
    expect(selector1).toHaveBeenCalledTimes(1)
    expect(ecosystem.dehydrate('@@selector')).toEqual({
      '@@selector-mockConstructor-0': 1,
    })

    act(() => {
      fireEvent.click(button)
      jest.runAllTimers()
    })

    expect(div.innerHTML).toBe('1')
    expect(selector1).toHaveBeenCalledTimes(2)
    expect(ecosystem.dehydrate('@@selector')).toEqual({
      '@@selector-mockConstructor-0': 1,
    })
  })

  test('argsComparator prevents a selector from running', async () => {
    jest.useFakeTimers()
    const atom1 = atom('1', (id: number) => id)
    const selector1 = {
      argsComparator: () => true,
      selector: jest.fn(({ get }: AtomGetters, id: number) => get(atom1, [id])),
    }

    function Test() {
      const [state, setState] = useState(0)
      const val = useAtomSelector(selector1, state)

      return (
        <>
          <button
            data-testid="button"
            onClick={() => setState(state => state + 1)}
          />
          <div data-testid="text">{val}</div>
        </>
      )
    }

    const { findByTestId } = renderInEcosystem(<Test />)

    const button = await findByTestId('button')
    const div = await findByTestId('text')

    expect(div.innerHTML).toBe('0')
    expect(selector1.selector).toHaveBeenCalledTimes(1)
    expect(ecosystem.dehydrate('@@selector')).toEqual({
      '@@selector-mockConstructor-0-[0]': 0,
    })

    act(() => {
      fireEvent.click(button)
      jest.runAllTimers()
    })

    expect(div.innerHTML).toBe('0')
    expect(selector1.selector).toHaveBeenCalledTimes(1)
    expect(ecosystem.dehydrate('@@selector')).toEqual({
      '@@selector-mockConstructor-0-[0]': 0,
    })
  })

  test('argsComparator does not prevent an inline selector from (re)-running, but the previous args are reused', async () => {
    jest.useFakeTimers()
    const atom1 = atom('1', (id: number) => id)
    const selector1 = jest.fn(({ get }: AtomGetters, id: number) =>
      get(atom1, [id])
    )

    function Test() {
      const [state, setState] = useState(0)
      const val = useAtomSelector(
        {
          argsComparator: () => true,
          selector: selector1,
        },
        state
      )

      return (
        <>
          <button
            data-testid="button"
            onClick={() => setState(state => state + 1)}
          />
          <div data-testid="text">{val}</div>
        </>
      )
    }

    const { findByTestId } = renderInEcosystem(<Test />)

    const button = await findByTestId('button')
    const div = await findByTestId('text')

    expect(div.innerHTML).toBe('0')
    expect(selector1).toHaveBeenCalledTimes(1)
    expect(ecosystem.dehydrate('@@selector')).toEqual({
      '@@selector-mockConstructor-0-[0]': 0,
    })

    act(() => {
      fireEvent.click(button)
      jest.runAllTimers()
    })

    expect(div.innerHTML).toBe('0')
    expect(selector1).toHaveBeenCalledTimes(2)
    expect(ecosystem.dehydrate('@@selector')).toEqual({
      '@@selector-mockConstructor-0-[0]': 0,
    })
  })

  test('force-destroying a selector after a render is scheduled results in one render', async () => {
    jest.useFakeTimers()
    const selector1 = jest.fn(() => 1)

    let renders = 0

    function Test() {
      const [, setState] = useState(0)
      const val = useAtomSelector({
        argsComparator: () => true,
        selector: selector1,
      })

      renders++

      return (
        <>
          <button
            data-testid="button"
            onClick={() => setState(state => state + 1)}
          />
          <div data-testid="text">{val}</div>
        </>
      )
    }

    const { findByTestId } = renderInEcosystem(<Test />)

    const button = await findByTestId('button')
    const div = await findByTestId('text')

    expect(div.innerHTML).toBe('1')
    expect(selector1).toHaveBeenCalledTimes(1)
    expect(ecosystem.dehydrate('@@selector')).toEqual({
      '@@selector-mockConstructor-0': 1,
    })
    expect(renders).toBe(1)

    act(() => {
      fireEvent.click(button)
      ecosystem.find('@@selector')?.destroy()
      jest.runAllTimers()
    })

    expect(div.innerHTML).toBe('1')
    expect(selector1).toHaveBeenCalledTimes(2)
    expect(ecosystem.dehydrate('@@selector')).toEqual({
      '@@selector-mockConstructor-0': 1,
    })
    expect(renders).toBe(2)
  })

  test('selector is not considered inline if it has multiple dependents', async () => {
    jest.useFakeTimers()
    const selector1 = jest.fn(() => 1)
    const selector2 = jest.fn(() => 2)

    function Test() {
      const [state, setState] = useState(0)
      const val = useAtomSelector(state % 2 ? selector1 : selector2)

      return (
        <>
          <button
            data-testid="button"
            onClick={() => setState(state => state + 1)}
          />
          <div data-testid="text">{val}</div>
        </>
      )
    }

    const instance = ecosystem.getNode(selector2)
    instance.on(() => {})

    const { findByTestId } = renderInEcosystem(<Test />)

    const button = await findByTestId('button')
    const div = await findByTestId('text')

    expect(div.innerHTML).toBe('2')
    expect(selector1).not.toHaveBeenCalled()
    expect(selector2).toHaveBeenCalledTimes(1)
    expect(ecosystem.dehydrate('@@selector')).toEqual({
      '@@selector-mockConstructor-0': 2,
    })

    act(() => {
      fireEvent.click(button)
      jest.runAllTimers()
    })

    expect(div.innerHTML).toBe('1')
    expect(selector1).toHaveBeenCalledTimes(1)
    expect(selector2).toHaveBeenCalledTimes(1)
    expect(ecosystem.dehydrate('@@selector')).toEqual({
      '@@selector-mockConstructor-0': 2,
      // id # 2 'cause `.on` generated id 1:
      '@@selector-mockConstructor-2': 1,
    })
  })

  test('inline selectors are swapped out in strict mode double-renders', async () => {
    jest.useFakeTimers()
    const atom1 = atom('1', () => ({ val: 1 }))

    function Test() {
      const [state, setState] = useState(0)
      const val = useAtomSelector(({ get }) => get(atom1).val)

      return (
        <>
          <button
            data-testid="button"
            onClick={() => setState(state => state + 1)}
          />
          <div data-testid="text">
            {val}
            {state}
          </div>
        </>
      )
    }

    const { findByTestId } = renderInEcosystem(<Test />, {
      useStrictMode: true,
    })

    const button = await findByTestId('button')
    const div = await findByTestId('text')

    snapshotSelectorNodes()
    expect(div.innerHTML).toBe('10')

    act(() => {
      fireEvent.click(button)
      jest.runAllTimers()
    })

    expect(div.innerHTML).toBe('11')

    snapshotSelectorNodes()
  })

  test('inline selector that returns a different object reference every time only triggers one extra rerender (strict mode off)', async () => {
    jest.useFakeTimers()
    const atom1 = atom('1', () => ({ val: 1 }))
    let renders = 0

    function Test() {
      renders++
      const { val } = useAtomSelector(({ get }) => ({ val: get(atom1).val }))

      return (
        <>
          <div data-testid="text">{val}</div>
        </>
      )
    }

    const { findByTestId } = renderInEcosystem(<Test />)

    const div = await findByTestId('text')

    expect(div.innerHTML).toBe('1')
    expect(renders).toBe(1)
    snapshotNodes()

    act(() => {
      ecosystem.getInstance(atom1).set({ val: 2 })
      jest.runAllTimers()
    })

    expect(div.innerHTML).toBe('2')
    expect(renders).toBe(2)
    snapshotNodes()
  })

  test('inline selector that returns a different object reference every time only triggers one extra rerender (strict mode on)', async () => {
    jest.useFakeTimers()
    const atom1 = atom('1', () => ({ val: 1 }))
    let renders = 0

    function Test() {
      renders++
      const { val } = useAtomSelector(({ get }) => ({ val: get(atom1).val }))

      return (
        <>
          <div data-testid="text">{val}</div>
        </>
      )
    }

    const { findByTestId } = renderInEcosystem(<Test />, {
      useStrictMode: true,
    })

    const div = await findByTestId('text')

    expect(div.innerHTML).toBe('1')
    expect(renders).toBe(4) // 2 rerenders + 2 for strict mode
    snapshotNodes()

    act(() => {
      ecosystem.getInstance(atom1).set({ val: 2 })
      jest.runAllTimers()
    })

    expect(div.innerHTML).toBe('2')
    expect(renders).toBe(6) // 3 rerenders + 3 for strict mode
    snapshotNodes()
  })

  test('inline selector stays subscribed after being swapped out', async () => {
    jest.useFakeTimers()
    const atom1 = atom('1', () => ({ val: 1 }))
    const selector = ({ get }: AtomGetters) => get(atom1)

    function Test() {
      const { val } = useAtomSelector({
        resultsComparator: (a, b) => a.val === b.val,
        selector,
      })

      return (
        <>
          <div data-testid="text">{val}</div>
        </>
      )
    }

    const { findByTestId } = renderInEcosystem(<Test />, {
      useStrictMode: true,
    })

    const div = await findByTestId('text')

    expect(div.innerHTML).toBe('1')

    act(() => {
      ecosystem.getInstance(atom1).set({ val: 2 })
      jest.runAllTimers()
    })

    expect(div.innerHTML).toBe('2')

    act(() => {
      ecosystem.getInstance(atom1).set({ val: 3 })
      jest.runAllTimers()
    })

    expect(div.innerHTML).toBe('3')

    act(() => {
      ecosystem.getInstance(atom1).set({ val: 4 })
      jest.runAllTimers()
    })

    expect(div.innerHTML).toBe('4')
  })

  test('when a selector is destroyed, other selectors that pass different params to the same ref retain their ids', async () => {
    jest.useFakeTimers()
    const atom1 = atom('1', () => '1')
    const selector1 = ({ get }: AtomGetters, str: string) => get(atom1) + str

    function Test() {
      const a = useAtomSelector(selector1, 'a')
      const b = useAtomSelector(selector1, 'b')
      const c = useAtomSelector(selector1, 'c')

      return (
        <div data-testid="text">
          {a}
          {b}
          {c}
        </div>
      )
    }

    const { findByTestId } = renderInEcosystem(<Test />)

    const div = await findByTestId('text')

    expect(div.innerHTML).toBe('1a1b1c')
    expect(ecosystem.dehydrate('@@selector')).toMatchInlineSnapshot(`
      {
        "@@selector-selector1-0-["a"]": "1a",
        "@@selector-selector1-0-["b"]": "1b",
        "@@selector-selector1-0-["c"]": "1c",
      }
    `)

    act(() => {
      ecosystem.getNode(selector1, ['a']).destroy(true)
      jest.runAllTimers()
    })

    expect(div.innerHTML).toBe('1a1b1c')
    expect(ecosystem.dehydrate('@@selector')).toMatchInlineSnapshot(`
      {
        "@@selector-selector1-0-["a"]": "1a",
        "@@selector-selector1-0-["b"]": "1b",
        "@@selector-selector1-0-["c"]": "1c",
      }
    `)
  })
})
