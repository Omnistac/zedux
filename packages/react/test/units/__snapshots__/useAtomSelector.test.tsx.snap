// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useAtomSelector inline selector that returns a different object reference every time only triggers one extra rerender (strict mode off) 1`] = `
{
  "1": {
    "className": "AtomInstance",
    "observers": {
      "@selector(unknown)-1": {
        "flags": 1,
        "operation": "get",
        "p": undefined,
      },
    },
    "sources": {},
    "state": {
      "val": 1,
    },
    "status": "Active",
    "weight": 1,
  },
  "@component(Test)-:ra:": {
    "className": "ExternalNode",
    "observers": {},
    "sources": {
      "@selector(unknown)-1": {
        "flags": 5,
        "operation": "useAtomSelector",
      },
    },
    "state": undefined,
    "status": "Active",
    "weight": 3,
  },
  "@selector(unknown)-1": {
    "className": "SelectorInstance",
    "observers": {
      "@component(Test)-:ra:": {
        "flags": 5,
        "operation": "useAtomSelector",
      },
    },
    "sources": {
      "1": {
        "flags": 1,
        "operation": "get",
        "p": undefined,
      },
    },
    "state": {
      "val": 1,
    },
    "status": "Active",
    "weight": 2,
  },
}
`;

exports[`useAtomSelector inline selector that returns a different object reference every time only triggers one extra rerender (strict mode off) 2`] = `
{
  "1": {
    "className": "AtomInstance",
    "observers": {
      "@selector(unknown)-1": {
        "flags": 1,
        "operation": "get",
        "p": undefined,
      },
    },
    "sources": {},
    "state": {
      "val": 2,
    },
    "status": "Active",
    "weight": 1,
  },
  "@component(Test)-:ra:": {
    "className": "ExternalNode",
    "observers": {},
    "sources": {
      "@selector(unknown)-1": {
        "flags": 5,
        "operation": "useAtomSelector",
      },
    },
    "state": undefined,
    "status": "Active",
    "weight": 3,
  },
  "@selector(unknown)-1": {
    "className": "SelectorInstance",
    "observers": {
      "@component(Test)-:ra:": {
        "flags": 5,
        "operation": "useAtomSelector",
      },
    },
    "sources": {
      "1": {
        "flags": 1,
        "operation": "get",
        "p": undefined,
      },
    },
    "state": {
      "val": 2,
    },
    "status": "Active",
    "weight": 2,
  },
}
`;

exports[`useAtomSelector inline selector that returns a different object reference every time only triggers one extra rerender (strict mode on) 1`] = `
{
  "1": {
    "className": "AtomInstance",
    "observers": {
      "@selector(unknown)-1": {
        "flags": 1,
        "operation": "get",
        "p": undefined,
      },
    },
    "sources": {},
    "state": {
      "val": 1,
    },
    "status": "Active",
    "weight": 1,
  },
  "@component(Test)-:rb:": {
    "className": "ExternalNode",
    "observers": {},
    "sources": {
      "@selector(unknown)-1": {
        "flags": 5,
        "operation": "useAtomValue",
      },
    },
    "state": undefined,
    "status": "Active",
    "weight": 3,
  },
  "@selector(unknown)-1": {
    "className": "SelectorInstance",
    "observers": {
      "@component(Test)-:rb:": {
        "flags": 5,
        "operation": "useAtomValue",
      },
    },
    "sources": {
      "1": {
        "flags": 1,
        "operation": "get",
        "p": undefined,
      },
    },
    "state": {
      "val": 1,
    },
    "status": "Active",
    "weight": 2,
  },
}
`;

exports[`useAtomSelector inline selector that returns a different object reference every time only triggers one extra rerender (strict mode on) 2`] = `
{
  "1": {
    "className": "AtomInstance",
    "observers": {
      "@selector(unknown)-1": {
        "flags": 1,
        "operation": "get",
        "p": undefined,
      },
    },
    "sources": {},
    "state": {
      "val": 2,
    },
    "status": "Active",
    "weight": 1,
  },
  "@component(Test)-:rb:": {
    "className": "ExternalNode",
    "observers": {},
    "sources": {
      "@selector(unknown)-1": {
        "flags": 5,
        "operation": "useAtomValue",
      },
    },
    "state": undefined,
    "status": "Active",
    "weight": 3,
  },
  "@selector(unknown)-1": {
    "className": "SelectorInstance",
    "observers": {
      "@component(Test)-:rb:": {
        "flags": 5,
        "operation": "useAtomValue",
      },
    },
    "sources": {
      "1": {
        "flags": 1,
        "operation": "get",
        "p": undefined,
      },
    },
    "state": {
      "val": 2,
    },
    "status": "Active",
    "weight": 2,
  },
}
`;

exports[`useAtomSelector inline selectors are swapped out in strict mode double-renders 1`] = `
{
  "@selector(unknown)-1": {
    "className": "SelectorInstance",
    "observers": {
      "@component(Test)-:r9:": {
        "flags": 5,
        "operation": "useAtomValue",
      },
    },
    "sources": {
      "1": {
        "flags": 1,
        "operation": "get",
        "p": undefined,
      },
    },
    "state": 1,
    "status": "Active",
    "weight": 2,
  },
}
`;

exports[`useAtomSelector inline selectors are swapped out in strict mode double-renders 2`] = `
{
  "@selector(unknown)-1": {
    "className": "SelectorInstance",
    "observers": {
      "@component(Test)-:r9:": {
        "flags": 5,
        "operation": "useAtomValue",
      },
    },
    "sources": {
      "1": {
        "flags": 1,
        "operation": "get",
        "p": undefined,
      },
    },
    "state": 1,
    "status": "Active",
    "weight": 2,
  },
}
`;
